### ISSUE TITLE
Add nullable column support

### ISSUE THREAD

**Issue by JayWhite2357** [spaceandtimelabs/sxt-proof-of-sql#183]
# Background and Motivation
Currently, we have not implemented support for columns with null values. This is a very useful and ubiquitous primitive for any database engine.

# Changes Required
- [ ] Support adding a flag to the columns types in order to support nullable columns. Logically, this is the same as adding a new column datatype for each existing datatype.
- [ ] Support existing operations on nullable columns. (e.g. we should be able to add a nullable bigint to a non-nullable bigint)

**Comment by JayWhite2357** [comment]
@iajoiner Can you scope out this issue a bit more and come up with a concrete plan.

Alternatively, if someone claims the bounty and is able to come up with a satisfactory plan themselves, I will tip an extra $1000 after completion.

**Comment by JayWhite2357** [comment]
@iajoiner Can you scope out this issue a bit more and come up with a concrete plan.

Alternatively, if someone claims the bounty and is able to come up with a satisfactory plan themselves, I will tip some extra toward.

**Comment by ssddOnTop** [comment]
@JayWhite2357 I am new to the project.. can you provide some sample commands or something to test against? Something that can help me to navigate through the code easily and more importantly, understand the issue properly.

**Comment by ssddOnTop** [comment]
Also, can someone please provide link to the discord server for easier communications

**Comment by JayWhite2357** [comment]
Take a look at the [examples](https://github.com/spaceandtimelabs/sxt-proof-of-sql/tree/main?tab=readme-ov-file#examples).
Also, take a look at the tests here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/tree/main/crates/proof-of-sql/tests.

**Comment by brainless** [comment]
@JayWhite2357 I am not planning to attempt this but I would happy to scope this out. I will look into your linked tests.

**Comment by JayWhite2357** [comment]
@brainless I can't promise a tip for just scoping it out. The bounty is specifically for completing the issue.
This is a potentially complex issue, where the correct, full design will likely not be clear until the code is complete.
To this point, @ssddOnTop, you may wish to get feedback from @iajoiner and myself on any design before you actually go through the effort of implementing it.

**Comment by brainless** [comment]
@JayWhite2357 sorry for the mis-communication. I did not mean a tip at all. I feel it is a great way for me to learn about the project before I want to take any bounty in the future.

**Comment by JayWhite2357** [comment]
> @JayWhite2357 sorry for the mis-communication. I did not mean a tip at all. I feel it is a great way for me to learn about the project before I want to take any bounty in the future.

No worries! I wasn't sure, and wanted to make sure expectations were set correctly. That sounds great!

**Comment by JayWhite2357** [comment]
/bounty $5000
Oops.

**Comment by ssddOnTop** [comment]
> To this point, @ssddOnTop (or anyone else wishing to attempt this), you may wish to get feedback from (avoiding ping) and myself on any design before you actually go through the effort of implementing it.

Sorry for delays.. I just moved in to a new place and I don't have access to my PC, and it's very difficult to do anything on arm based mac and docker.

I'll definitely start making some progress by tomorrow.

Meanwhile @JayWhite2357 can you provide a link to the discord channel for easier communication?

**Comment by ssddOnTop** [comment]
Is there any way to run this on cpu? @JayWhite2357

**Comment by iajoiner** [comment]
Here are some basic thoughts.
- This issue can be resolved in very complex and unmaintainable manners or it can be resolved in simple ones. We strongly prefer the latter.
- A nullable column, `c`, is basically two different non-nullable columns, the values column `c_v` and the presence column `c_p` which is Boolean
- Any unary operation on a nullable column `c` is really just the operation on `c_v` while the presence column of the result is simply `c_p`.
- Similarly for binary operations with at least one side non-nullable we can trivially reuse results from the same binary operation on non-nullable columns.
- For binary operations with both sides `a` and `b` nullable we are just applying the operation on `a_v` and `b_v` and at the same time apply AND on `a_p` and `b_p`. Since AND is already provable we can just reuse the result.

**Comment by JayWhite2357** [comment]
> Is there any way to run this on cpu? @JayWhite2357

Yes. Either set the environment variable `BLITZAR_BACKEND=cpu`, which still requires linux. Or, disable the "blitzar" features, which should run on any os.

**Comment by iajoiner** [comment]
This issue should be split into multiple PRs. I ordered the tasks so that tasks below are not dependent on tasks above and the code with incremental changes should still compile and run properly with CI passing.
- Allow nullability for columns in `base/database`.
  -  Add `NullableColumn` that contains a `values` `Column`, and an optional boolean `presence` slice. `presence` being `None` indicates that the column is not nullable.
  -  Add `NullableOwnedColumn` that contains a `values` `OwnedColumn` and an optional boolean `presence` vector.  `presence` being `None`  indicates that the column is not nullable.
      - It might be a good idea to rename `NullableColumn` to `Column` and the current `Column` to `NonnullableColumn` etc. Make sure to make it a breaking change.
  - Implement [column operations](https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql/src/base/database/column_operation.rs) and [expression evaluation](https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql/src/base/database/expression_evaluation.rs) for nullable columns. Note that you should reuse the code for non-nullable ones and follow the guidelines above. Just use a `match` and your work will be mostly reduced to doing an AND on presence columns if both sides are nullable.
  - Extend Arrow conversions to nullable columns.
- Allow nullability for tables in `base/database`.
  - Make columns in `OwnedTable` nullable.
  - Make sure Arrow conversions and utils still work after the change.
- Allow nullability in `sql/proof_exprs` and `sql/proof_plans`.
  - Implement `IS NULL`, `IS NOT NULL` which is basically an `EqualExpr` on presence columns.
  - Implement `IS TRUE` which means `is not false and not null` which we need for `WHERE` clauses.
  - Extend proofs for unary and binary operations to nullable columns. Please do so with minimal changes.
    - It might be a good idea to refactor the code and make all binary operations use a generic `BinaryOperationExpr` and add nullability to it. Same for unary operations (just `NOT` for now, more to come).
  - None of the `ProofPlan`s should require serious changes to be nullable other than accommodation of `ProofExpr` changes.
    - If the workload here is particularly high please split changes to `ProofPlan`.
- Allow nullability in `sql/postprocessing`
  - Most of the work should already be complete when you do the work for columns. Do make sure test utilities etc are updated correctly.
- Allow nullability in `sql/parse`.
  - Use implicit `IS TRUE` in `WHERE` clauses.
  - Get everything to work. This most likely consists of lots of miscellaneous tasks.
  - Add additional tests for tables with nullability [here](https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql/src/sql/parse/query_expr_tests.rs).
    - After this the task `proof-of-sql` crate is done.
- Allow nullability in `proof-of-sql-parser`
  - Make sure to do this task AFTER the work for `proof-of-sql` is complete since otherwise it will break the `proof-of-sql` crate.
  - Make [sql.lalrpop](https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql-parser/src/sql.lalrpop) accept nulls.
  - Make corresponding changes in [intermediate_ast.rs](https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql-parser/src/intermediate_ast.rs) and elsewhere. Get the tests to run.
  - Add [integration tests](https://github.com/spaceandtimelabs/sxt-proof-of-sql/tree/main/crates/proof-of-sql/tests) for nullable tables using both `InnerProductProof` and `DoryEvaluationProof`.

**Comment by ssddOnTop** [comment]
why do we have `Serialize, Deserialize` on `ColumnType`? Isn't that an internal struct?

I am confused on the flow of code

**Comment by ssddOnTop** [comment]
why do we have (De)serialize on ColumnType? Isn't that an internal struct?

I am confused on the flow of code

@iajoiner @JayWhite2357

**Comment by digital-phoenix** [comment]
@iajoiner what should the result of a binary operation be when one of the values is null?

**Comment by JayWhite2357** [comment]
> @iajoiner what should the result of a binary operation be when one of the values is null?

In general, we should aim to be as similar to PostgreSQL as possible.

So, `1+NULL` should be `NULL`. However, bool logic is not as straightforward in PostgreSQL (and most engines): https://www.postgresql.org/docs/current/functions-logical.html#FUNCTIONS-LOGICAL.

**Comment by maujim** [comment]
Do you have a link to the Discord @JayWhite2357? I'd like to join and work on this bounty

**Comment by iajoiner** [comment]
@maujim Sorry for my late reply. Here we go: https://discord.gg/qnKtErW3

**Comment by JayWhite2357** [comment]
FYI: increased this bounty amount.

Added an additional note to the issue.

**Comment by ntoledo319** [comment]
PR updated: claim line added and PR body sanitized for hidden unicode. Video demo placeholder posted with commands to run nullable column proofs; please attach the recorded `demo_small.mp4` to the PR comment when available.

**Comment by rafael81** [comment]
/attempt #183

Submitted PR #1411: https://github.com/spaceandtimefdn/sxt-proof-of-sql/pull/1411

This is intentionally a small partial improvement rather than a full nullable-column implementation. It adds a uniform guard so Arrow arrays with nulls are rejected before conversion into non-nullable `OwnedColumn` values, preventing null slots from being silently interpreted through Arrow backing values while explicit nullable support is still absent.

Validation listed in the PR body:
- `cargo fmt --check`
- focused nullable Arrow regression test
- owned/arrow conversion tests
- `git diff --check`
- `source scripts/check_commits.sh`

Implemented with AI assistance, then manually reviewed and tested.

**Comment by zhaog100** [comment]
👋 I'd like to work on this bounty. Will start with detailed analysis and implementation plan.

**Comment by ncthuc2004** [comment]
Hi maintainers, I can take this issue and submit a clean PR quickly.

- Stack fit: React/Next.js/TypeScript/Java automation
- ETA: 2-8 hours after confirmation
- Delivery: code + tests + validation notes

If this bounty is open for external contributors, please confirm payout via https://paypal.me/ThucNguyen2004.

**Comment by mt211211** [comment]
Opened PR #1955 for this bounty: https://github.com/spaceandtimefdn/sxt-proof-of-sql/pull/1955

This is a focused nullable-column proof of concept rather than a metadata-only change. It includes nullable schema propagation, canonical validity-mask utilities, nullable owned/borrowed column wrappers, nullable `BigInt` addition paths, and proof coverage that filters nullable rows through an explicit validity column before proving projection results.

Validation passed locally:

```powershell
$env:CARGO_TARGET_DIR='C:\sxt-target'; $env:CARGO_BUILD_JOBS='1'; cargo test -p proof-of-sql nullable --no-default-features --features "arrow cpu-perf test"
$env:CARGO_TARGET_DIR='C:\sxt-target'; $env:CARGO_BUILD_JOBS='1'; cargo test -p proof-of-sql validity --no-default-features --features "arrow cpu-perf test"
$env:CARGO_TARGET_DIR='C:\sxt-target'; $env:CARGO_BUILD_JOBS='1'; cargo test -p proof-of-sql-planner we_can_convert_column_fields_to_schema --features cpu-perf
```

**Comment by Devpen787** [comment]
Hi maintainers. I started a small local POC for nullable input columns and noticed the recent broader POC in #1955, so I want to avoid duplicating work.

My local slice is intentionally narrow:

- keep existing `OwnedColumn` behavior stable
- represent nullable data as a value column plus explicit validity mask
- avoid silently treating `NULL` as `0`
- show nullable `BIGINT` as value + validity commitments
- add nullable Arrow `Int64Array` ingestion for one primitive path
- split nullable input into existing engine-compatible value and validity `OwnedColumn`s
- lift those columns into an `OwnedTable` while preserving row alignment

Focused local validation is passing:

```bash
cargo test -p proof-of-sql nullable --no-default-features --features="arrow rayon ark-ec/parallel ark-poly/parallel ark-ff/asm"
```

Given #1955 is now open, would it be more useful for me to:

1. help test/review that PR against a smaller macOS/local CPU path,
2. turn my narrow slice into a small complementary PR if there is a gap,
3. or pick up the next nullable work item after the maintainers decide on #1955?

I do not want to add noise or compete with an overlapping implementation if review/testing or a smaller follow-up would be more useful.

**Comment by MyTH-zyxeon** [comment]
/attempt #183

I'd like to work on this bounty with a complementary, low-overlap path.

Context: I already have recent contributions in this repo through PRs #1771, #1774, and #1820, so I can iterate in the project conventions and test surface quickly. I also see the broader nullable PoC in #1955, so I want to avoid duplicating that work unless maintainers explicitly want a competing implementation.

## Proposed scope
Start with a review/test/follow-up slice that helps maintainers decide or land the nullable work:

1. Review #1955 against the original maintainer guidance in this issue, especially the `(values, presence)` model, Arrow ingestion, row-alignment invariants, and null propagation through proof expressions.
2. Add or propose a focused complementary test path if there is a missing acceptance slice, preferably a macOS/local CPU-friendly path that validates nullable Arrow `Int64Array` ingestion, validity-mask alignment, and proof verification without silently treating `NULL` as `0`.
3. If #1955 needs a smaller follow-up PR rather than review-only help, prepare a narrow PR for the missing slice instead of reimplementing the whole nullable design.

## Deliverables
- A concise review/test report against #1955.
- A small PR only if the report identifies a concrete gap that can be fixed independently.
- Validation notes using the repo's existing Rust test commands and no live service/secrets dependency.

## ETA
1 day for the review/test report, then 2-3 days for a small complementary PR if maintainers prefer that route.

Before I start coding: should I focus on review/testing for #1955 first, or open a separate narrow follow-up PR for the first missing nullable acceptance slice?

**Comment by MyTH-zyxeon** [comment]
# Paste-ready comment (Issue #183: Add nullable column support)

Context: https://github.com/spaceandtimefdn/sxt-proof-of-sql/issues/183 (Bounty)

---

Hi SxT team — I’d like to propose a small, reviewable PoC plan before any large nullable-column PR.

## PoC goal (per issue request)
Demonstrate at least one non-trivial proof that includes at least one nullable type (i.e., nulls actually flow through constraints and the proof).

## Proposed representation
- Add a nullability marker at the column type layer (either `Nullable(T)` wrapper or `is_nullable` flag + a `NullBitmap`).
- Treat nulls as: `is_null[row] == 1` implies the value is ignored by semantics; operations produce appropriate null-ness (SQL-ish `NULL` propagation) or a clearly documented subset for PoC.

## Minimal scope for PoC PR
1) **Data model**: introduce `NullBitmap` + wire it through the column/field metadata.
2) **One datatype first**: `Nullable<BigInt>` (or whichever is easiest in the codebase).
3) **One operation**: support `+` between `Nullable<BigInt>` and `BigInt` with null propagation.
4) **One end-to-end proof test**: a query/proof that uses nullable values and asserts correct output + verifier success.

## Follow-up slices (only after PoC acceptance)
- Extend to more types and operations.
- Add Arrow ingestion/serialization support for nullable columns (if desired by maintainers).

## Quick questions to avoid rework
- Preferred null semantics for PoC: strict SQL 3-valued logic, or a simpler “null-propagates in arithmetic/comparison” subset?
- Do you want `Nullable(T)` as a distinct type, or `T + bitmap` as orthogonal metadata?

If this plan aligns, I can prepare the PoC PR in small commits so each step stays reviewable.

**Comment by MyTH-zyxeon** [comment]
Nullable column support (#183) – focused PoC proposal (before any large PR)

Motivation
- I saw the note requesting a Proof of Concept that demonstrates a non-trivial proof involving at least one null type.
- Below is a concrete, minimal PoC plan that tries to (1) keep surface area tight, (2) exercise both the data model and proof path, and (3) leave room for an eventual full null-support rollout.

Questions for maintainers (to avoid wrong-direction work)
1) Null semantics: should this follow SQL 3-valued logic (NULL propagation) for arithmetic + comparisons, or a simplified “unknown propagates” rule set for v1?
2) Preferred representation: validity bitmap (Arrow-style) + values column, or “Option<T> per row” encoded differently?
3) Is there an expected query/proof target you want the PoC to cover (e.g., filter + aggregation, join, group-by), or is a small arithmetic + filter proof acceptable?

Proposed PoC scope (1 nullable type, 1–2 operations, 1 proof)
A) Data model (minimal)
- Introduce a Nullable wrapper for one existing column type (suggest: BigInt or Scalar) as:
  - values: existing non-nullable column representation
  - validity: bitmask/bitmap with 1=present, 0=null
- Add a single new column variant (e.g., NullableBigInt/NullableScalar) rather than refactoring every column type in PoC.

B) Semantics (minimal but testable)
- For binary op `add(nullable, non-nullable)` and `add(nullable, nullable)`:
  - if either operand is null => result null
  - else normal addition
- For filter predicate `is_not_null(nullable_col)` to allow proving “null handling is correct” with both kept and dropped rows.

C) Proof target (non-trivial but narrow)
- Add a proofable query that combines null-propagation + filtering so the proof has to “carry” validity information:
  - Example: SELECT SUM(a + b) FROM t WHERE is_not_null(a) AND is_not_null(b)
  - With `a` nullable, `b` non-nullable (or both nullable) so we hit both null and non-null cases.
- Add a fixed test fixture with mixed null/non-null rows and assert the public output matches the expected SQL semantics.

D) PR structure suggestion (to keep reviewable)
1) Nullable column representation + (de)serialization invariants + unit tests
2) Implement `is_not_null` + `add` for nullable column(s) + unit tests
3) End-to-end proof test for the chosen query (the PoC acceptance gate)

Acceptance criteria (for PoC)
- Clear docs/commented invariants for validity bitmap length and alignment
- Unit tests that cover: all-null, none-null, alternating nulls, and null propagation
- One end-to-end proof test that fails if validity is ignored/mishandled

If this direction matches what you want, I can follow up with a PoC-only PR (single nullable type first) and keep the implementation isolated so it doesn’t lock you into a full nullable rollout.

**Comment by MyTH-zyxeon** [comment]
## Draft: proposal / design questions for sxt-proof-of-sql#183 (Nullable column support)

Goal: clarify acceptance criteria + land an implementation plan that is easy to review.

### Questions for maintainers
1) **Semantics**: should NULL be modeled as an explicit value in the commitment/PCS layer, or as an *absent* value (e.g., extra selector + placeholder scalar)?
2) **Encoding**: do you prefer a canonical encoding (e.g., `Option<T>` → `(is_null, value)`), and if so which field element is used for NULL sentinel (if any)?
3) **Scope**: which column types must support nullable first (scalars only vs. byte arrays, decimals, timestamps)?
4) **Constraints**: any constraints from the prover/verifier or from SQL type mapping that should shape the API?
5) **Compatibility**: should this be a strictly additive API, or is a breaking change acceptable behind a feature flag?

### Proposed approach (additive, test-first)
- Introduce a `NullableColumn<T>` (or `CommittableColumn::Nullable(Box<CommittableColumn>)`) representation that carries both the value column and a `is_null` boolean column.
- Commitment generation: commit `value` and `is_null` in a deterministic order; document ordering so circuit/proof consumers match.
- Evaluation rules:
  - For expressions, propagate `is_null` through operations using SQL three-valued logic where relevant.
  - For comparisons, define explicit behavior (e.g., `NULL = NULL` is NULL/false depending on query semantics), and mirror whatever the engine expects.
- Serialization: extend the Dory / dynamic commitment setup to include the new `is_null` column; ensure stable ordering.
- Tests:
  - Golden tests for commitment determinism with mixed null/non-null rows.
  - Property tests around `is_null` propagation.
  - Regression tests for existing non-null columns (must remain identical commitments).

### Deliverable plan
- PR1: data model + commitment ordering + determinism tests.
- PR2: expression/eval semantics + additional integration tests.

If you can confirm preferred encoding and ordering constraints, I can tailor the PR shape to reduce review load.

**Comment by MyTH-zyxeon** [comment]
## Proposal: Proof-of-Concept plan for nullable column support (Issue #183)

I can start with a focused PoC that demonstrates **one nullable primitive end-to-end** (including a non-trivial proof), while keeping the surface area small and reviewable.

### 1) Representation
- Add a nullable wrapper for columns that carries:
  - `values: Column<T>` (existing representation)
  - `validity: BitVec/Bitmap` (1 = non-null, 0 = null)
- Commitment / proof inputs commit to both `values` and `validity` so nullness is part of the statement.

### 2) Minimal semantics for PoC
Implement only what’s needed for the PoC query:
- `IS NULL` / `IS NOT NULL`
- `COALESCE(nullable_i64, 0)` (or `IFNULL`) for one numeric type
- arithmetic `+` between `nullable_i64` and non-nullable `i64` (propagate nulls unless coalesced)

### 3) Concrete PoC query + proof
Use a query that forces both null tracking and computation:

```sql
SELECT SUM(COALESCE(a, 0))
FROM t
WHERE b IS NULL OR b = TRUE;
```

Where:
- `a` is `NULLABLE INT` (or bigint)
- `b` is `NULLABLE BOOLEAN`

This exercises:
- validity bitmap constraints
- boolean filtering with null-aware predicate
- coalesce + aggregation

### 4) Scope boundaries (to keep it mergeable)
- No attempt to generalize all operations/types in the first PR.
- One nullable numeric type + one nullable boolean type for the PoC.
- Add targeted tests + bench guardrails to ensure the added constraints don’t explode prover time.

### 5) Questions for maintainers (to avoid wasted work)
1) Preferred null encoding: explicit validity bitmap vs sentinel-value encoding?
2) For PoC, is `COALESCE` acceptable, or would you prefer `IS NULL` gating only?
3) Which proof path should the PoC use (planner/executor path vs a more direct proof_primitive path) to be maximally convincing for review?

If this plan matches expectations, I’ll start by landing the representation + validity constraints and a minimal query proof, then expand operation/type coverage incrementally.

**Comment by MyTH-zyxeon** [comment]
/attempt #183

Implementation plan (PoC first, then incremental expansion):

1) Scope & design review (0.5 day)
- Read current column type/encoding + commitment/proof plumbing.
- Decide representation for nullability: `NullBitmap` (bitset) + `Vec<T>` payload, wrapped as a new column variant (preferred) or a generic wrapper.

2) Minimal nullable column representation (1 day)
- Add a new nullable column abstraction that can be carried through existing evaluation/proof code without eagerly materializing.
- Provide conversions between nullable/non-nullable where required.

3) PoC semantics + one non-trivial proof (2–3 days)
- Implement null propagation semantics for at least:
  - one numeric type (e.g. BigInt/Int)
  - one operation (e.g. add) and one comparison/filter
- Ensure mixed nullable/non-nullable works (nullable + non-nullable, etc.).
- Add a PoC test proving a query involving nulls (as requested in the issue).

4) Hardening + docs/tests (1–2 days)
- Add property tests / targeted regression tests for null edge cases (all nulls, no nulls, mixed, offset windows).
- Document invariants (bitmap length, alignment, deterministic encoding).

Deliverable: a PoC PR demonstrating end-to-end null support on a small slice (one type + a couple ops) with a non-trivial proof, followed by incremental PRs to expand to more types/ops.

Time estimate: ~5–7 working days for the PoC PR.

**Comment by MyTH-zyxeon** [comment]
Status refresh (2026-06-02 UTC): still interested in landing a clean, reviewable nullable-column support slice for #183.

What I can do next (once I have a clean checkout + a reproducible failing scenario):
- Confirm acceptance criteria: which SQL surface(s) must accept NULLs (projection/filter/joins/group-by/aggregates) and which proof boundary must preserve NULL semantics.
- Add one minimal end-to-end regression test that demonstrates current failure (nullable Arrow ingestion -> query -> proof verify) and locks expected behavior.
- Implement the smallest vertical slice needed to make that test pass (likely: Arrow -> ColumnField/ColumnType -> planner/executor propagation + serialization/round-trip).

Blocking question: do maintainers have a recommended local repro command / test harness for nullable columns (e.g., a single `cargo test ...` target and a tiny Arrow batch fixture)?

If you point me at the preferred test entrypoint + expected NULL behavior, I can respond with a concrete PR plan + time estimate on top of the earlier PoC notes.

**Comment by MyTH-zyxeon** [comment]
Status refresh for bounty #183 (2026-06-02 UTC): there is now an upstream-visible landing path that looks narrower than starting a fresh competing implementation.

Current public state I can verify:
- PR #1955 is still OPEN, MERGEABLE, and only visibly BLOCKED by REVIEW_REQUIRED from the public PR metadata.
- A small stacked PR was posted publicly here: https://github.com/mt211211/sxt-proof-of-sql/pull/1
- That stacked PR claims to cover two of the highest-leverage acceptance gaps that were still discussed on #1955:
  1. direct nullable Arrow Int64Array ingestion preserving validity metadata
  2. ColumnField legacy/new serde coverage plus Arrow Field nullable-bit preservation

If that direction is acceptable, the lowest-risk landing path seems to be:
1. port/cherry-pick the focused tests/helpers from the stacked PR onto #1955,
2. rerun the nullable-focused test slice already documented there,
3. treat the planner/proof-boundary evidence as the next follow-up rather than broadening this PoC all at once.

That would turn the current local/stacked work into upstream-visible evidence without creating another competing #183 claim path.

If maintainers prefer a different acceptance order for #183, I can reframe the checklist around that.

**Comment by catcherintheroad-hub** [comment]
Small status update for the nullable-column bounty landing path:

- I reviewed #1955 and the current stacked PR from the acceptance-gap angle.
- I found one concrete Arrow correctness gap in the stacked nullable `Int64Array` conversion: raw `array.values().to_vec()` can ignore a sliced array's logical offset.
- I opened a small stacked follow-up PR against the existing nullable branch: https://github.com/mt211211/sxt-proof-of-sql/pull/2
- Upstream context/validation comment: https://github.com/spaceandtimefdn/sxt-proof-of-sql/pull/1955#issuecomment-4633524460

Validation passed locally with the CPU-friendly test slice:

```bash
cargo test -p proof-of-sql nullable --no-default-features --features "arrow test" -- --nocapture
```

Result: 10 nullable tests passed, including the new sliced nullable Arrow ingestion regression.
